fronts-client/src/components/FrontsEdit/FrontContent.tsx

```diff
@@ -94,6 +94,113 @@ interface FrontState {
 	isCollectionsStale: boolean | undefined;
 }
 
+function getNextGroupTarget(
+	currentGroupId: string,
+	groupIds?: string[],
+	groupsData?: Group[],
+) {
+	if (!groupIds || !groupsData) {
+		return;
+	}
+	const currentIndex = groupIds.findIndex((id) => id === currentGroupId);
+	const nextGroupId = groupIds[currentIndex + 1];
+	const nextGroup = groupsData.find((group) => group.uuid === nextGroupId);
+
+	return {
+		index: 0,
+		id: nextGroupId,
+		type: 'group',
+		groupIds: groupIds,
+		groupMaxItems: nextGroup?.maxItems,
+		groupsData: groupsData,
+		cards: nextGroup?.cardsData,
+	};
+}
+type CollectionMove<T> = {
+	to: PosSpec;
+	data: T;
+	from: null | PosSpec;
+	type: 'collection';
+};
+
+/**
+ * The move queue handles cascading card moves triggered when attempting to insert a card into a full group.
+ * It ensures the affected cards are shifted into subsequent groups, maintaining maxItems constraints.
+ */
+export const buildMoveQueue = (move: Move<TCard>) => {
+	const queue: CollectionMove<TCard>[] = [];
+
+	const { data: movedCard, to, from } = move;
+
+	// Determine if the card is being added to the end of a group.
+	const isBottomInsert = to.index === (to.cards?.length ?? 0);
+
+	// If inserting at the bottom of a full group, move the card to the next group instead.
+	const target = isBottomInsert
+		? getNextGroupTarget(to.id, to.groupIds, to.groupsData)
+		: to;
+
+	if (!target) return queue;
+	// Add the initial move: either to the original target or the adjusted group.
+	queue.push({
+		to: target,
+		data: movedCard,
+		from: from || null,
+		type: 'collection',
+	});
+
+	/**
+	 *  Begin cascading shifts for full groups.
+	 *  Start at the target group (or one after, if we adjusted for a bottom insert),
+	 *  and continue shifting the last card of each full group into the next group.
+	 * */
+
+	const startIndex = to.groupIds?.indexOf(to.id) ?? 0;
+
+	for (
+		let index = isBottomInsert ? startIndex + 1 : startIndex;
+		to.groupsData && index < to.groupsData.length - 1;
```

Comment:
Does this `-1` offset correspond to the potential `+1` in the line above?

Reply:
Not quite, it was there to avoid the cascading logic to trigger when we've reached the last group - but we've refactored to make it more explicit.